src/dune_rules/lib.ml

```diff
@@ -256,6 +256,7 @@ module T = struct
          {[ This kind of expression is not allowed as right-hand side of `let
          rec' }] *)
       mutable sub_systems : Sub_system0.Instance.t Lazy.t Sub_system_name.Map.t
+    ; modules : Modules.t Lazy.t option
```

Comment:
Shouldn't this be a `Memo.Lazy.t`? Using `Lazy.t` here seems dodgy

Reply:
Related remark: this field feels like it adds quite a bit of weight to the database. This might be an issue when we want the fast polling mode to share values between builds as we might have to hash these values. It's a bit sad given that this is only used for the "root module" feature, which is not yet widespread.

Another approach would be to add a memoized function with this API: `val modules_of_lib : Lib.Id.t -> Modules.t`. It might also avoid a few `Fdecl.t`. WDYT?

Reply:
> Shouldn't this be a Memo.Lazy.t? Using Lazy.t here seems dodgy

We're using `Lazy.t` for a few other fields in `Lib.t`. Should I change those as well?

> This might be an issue when we want the fast polling mode to share values between builds as we might have to hash these values. It's a bit sad given that this is only used for the "root module" feature, which is not yet widespread.

Only because we've been going through  contortions to avoid using it. It's quite useful actually, and I can use it to simplify the virtual libraries rules as well.

> Another approach would be to add a memoized function with this API: val modules_of_lib : Lib.Id.t -> Modules.t. It might also avoid a few Fdecl.t. WDYT?

Sure, I could give it a try. But I think we're going to need some sort of a database value in here. So it would be:

```
module DB : sig
  type t
  val create : Super_context.t -> t
end
val modules_of_lib : DB.t -> Lib.Id.t -> Modules.t
```

The DB would then be a field in super_context, and that would mean that Fdecl will be necessary after all.

Reply:
> We're using Lazy.t for a few other fields in Lib.t. Should I change those as well?

We'll need to look into these eventually, but not now. Let's leave it as it for now then.

> Only because we've been going through contortions to avoid using it. It's quite useful actually, and I can use it to simplify the virtual libraries rules as well.

Ack. In this case it seems more useful. I retract my remark. We might still want to explode such records in the future, but we can cross that bridge when we have a better idea of the performance bottlenecks.

> Sure, I could give it a try. But I think we're going to need some sort of a database value in here. So it would be:

Doesn't seem worth it, let's leave it as it for now.